Fix: create dCampaign search page + navbar search box integration

khadis123 opened the pull request:
Dynamic Search:

1.	Navbar SearchBox

- User types a query → submits → redirected to /campaigns/search?q=....
- Input is cleared immediately after redirect to avoid confusion with CampaignSearchPage.

2.	CampaignSearchPage (http://localhost:3000/campaigns/search)

- Reads q, sort, category from URL parameters.
- Fetches campaigns accordingly and paginates results.
- Main input field is focused on mount, with cursor at the end of query for easy editing.
- Supports sorting, category filtering, pagination, and clearing filters.

khadis123 committed:
Fix: create Campaign search page + navbar search box integration

MounirAia commented on package.json:
```diff
@@ -46,7 +46,7 @@
     "fontkit": "^2.0.4",
     "input-otp": "^1.4.2",
     "lucide-react": "^0.542.0",
-    "next": "15.5.2",
```
Let’s avoid changing the Next.js version for now. What was the reason for updating it?

khadis123 replied:
I think it automatically updated when I run something. I will reverse it as it's not required for this functionality to work properly

MounirAia commented on tsconfig.json:
Hey! I noticed there’s a change to the tsconfig file here. Could you share what motivated this update?

khadis123 replied:
In dynamic-search branch uses "jsx": "react-jsx" — this enables the automatic JSX runtime. But as we are using Next.js probably it's better to keep it as it was before on master branch (uses "jsx": "preserve"). It was working with this setup but let's keep it as it was before as it works that way as well

MounirAia commented on app/api/campaigns/search/route.ts:
Could you refactor this so the route simply calls a separate server-side function? The idea is to move the search logic into a reusable function that receives the parameters (q, sort, category), and then have the GET handler just call that function.

You can take example of functions that are the `src/api/` folder.

MounirAia commented on src/lib/supabase/server.ts:
```diff
@@ -109,3 +109,5 @@ export const getOrganizationProfile = cache(
     return { email: user.email, ...profile };
   }
 );
+export { createClient };
```
I think we can remove this line, I don't see it being used in the code elsewhere

khadis123 replied:
I think it was just when I was testing it. Probably forgot to remove after testing. Will remove it

khadis123 committed:
Fix campaign search using search_campaigns function

khadis123 commented:
Thanks for the review!

I refactored the search implementation to follow the project’s existing patterns:
1. Introduced a new reusable search function:
Added src/api/campaigns.ts
This mirrors existing API patterns used in:
src/api/donations.ts
src/api/project.ts

2. Removed the old /api/campaigns/search route handler
The server route was no longer needed.
The frontend now calls the Supabase RPC function (search_campaigns) directly via the new server-side API helper.

3. Cleaned up previous changes
No more accidental modifications to tsconfig.json
No unintended Next.js config edits
Everything matches the master branch’s structure and conventions.

4. Manually tested locally
Confirmed correct behavior in front-end search UI.
You can run and test it to see if it works as expected. 
Let me know if everything ok now.

MounirAia commented on app/api/campaigns/search/route.ts:
I think this route handler isn’t needed anymore. Since you already have a **server function (`searchCampaigns`) that contains all the actual logic**, you can call that function directly in your **Server Component** instead of going through an API route.

If you remove this route file, the flow becomes simpler:

* Do the **fetch/search logic directly in the Server Component** (just call `searchCampaigns()` there).
* Use a `<Link>` to navigate to your `/search` page (or whatever route triggers the search UI).
* That page can then call the server function and render the results without needing an API round-trip.

In other words:
**No need for `/api/campaigns/search` since the logic already lives in a reusable server function — just import and call it directly.**

This keeps things fully server-side, reduces overhead, and aligns better with Next.js App Router patterns.

khadis123 replied:
If I do that then all the features of dynamic search etc won't work properly:

current Search page is fully dynamic:
typing in search → runs immediately (debounced)
pagination → client state
category filtering → client state
input auto-focusing
loading spinner shown while fetching
smooth transitions without page reload
All of this is possible only because the page is a Client Component.

Btw, there's another option - Hybrid approach:
Search page stays server component.
A small separate Client Component handles UI (search input only).
But the main search results come from server.
The page will fetch results from the server via a server function, not an API route.

I personally like to leave these features as they currently are because they work as intended. 
If it's critical to do it as you indicated above, I can do that but I guess all those features will be gone then. Do you want me to refactor everything anyway, or a hybrid option, or you have other ideas about it?

MounirAia replied:
Thank you for your comment, here is my review:

Yes, I prefer that you use a server function to fetch the data, rather than an API route, good idea.

Also:
- I did not see the loading spinner while fetching, are you referring to the text `Loading campaigns`
- When fetching campaigns with the search, the results returned does not display the image of the campaign
- What do you mean by input autofocus?
- When typing the search is not debounced, for every character's types there is a network request made

https://github.com/user-attachments/assets/62b30cd0-931f-479e-8e82-90d821accac9

MounirAia commented on package-lock.json:
the package lock json file still has the nextjs 16 version. Run `npm install` on your terminal to fix this.
Also to uninstall npm package it is better to use `npm uninstall <package-name>` rather than deleting the record on the package.json file

khadis123 committed:
Refactor campaign search to server-only with FTS and project area filtering

khadis123 commented:
This PR refactors the campaign search flow to be server-only, URL-driven, and scalable, while adding project area filtering and improving overall architecture consistency. 

What was done:
1. Server-only campaign search •	Removed API route indirection (app/api/campaigns/search/route.ts) 
•	Search is now handled directly via a server function: o	src/api/campaigns.ts → searchCampaigns() 
•	Page components call server logic directly (Next.js App Router best practice) 

2. Full-Text Search (FTS) 
•	Integrated Postgres Full-Text Search using search_vector 
•	Added migration to support indexed FTS 
•	Enables scalable keyword search without client-side logic 

3. Project Area filtering 
•	Wired project_areas dropdown to Supabase 
•	Supports: o	“All categories” default o	URL param: ?area=<id> 
•	Correctly filters campaigns via organization_project_areas 
•	Alphabetically sorted categories 

4. Minimal, GET-based search UI 
•	Uses pure GET form (no JS handlers) 
•	Search, sort, filter, pagination all driven by URL 
•	Autofocus on search input 
•	Clear button resets all filters correctly 

5. Consistent campaign cards 
•	Search results reuse the same ProjectCard as the main campaigns page 
•	Correct URLs: /campaigns/[orgSlug]/[projectSlug] 

6. Cleanup & structure 
•	Removed obsolete API route 
•	Added missing LoadingSpinner 
•	Improved typing and error handling 
•	Maintains server-only boundaries  

Why this approach: 
•	Scales well for large datasets (FTS + indexed queries) 
•	SEO-friendly and shareable URLs 
•	Decided to get rid of client state or debouncing as we are using mostly SSR approach 
•	Matches Next.js App Router & Supabase best practices 

Final notes:
•	Why no debounce? → Server FTS + GET search is intentional and scalable 
•	Why no API route? → App Router server functions are the recommended approach as I understand it
•	FTS + indexed queries scale better than client filtering
